### DISCUSSION TITLE
modify the response before returning it to the client: adding a virtual field

### DISCUSSION THREAD

**Discussion by danielres** [pocketbase/pocketbase#3820] [Q&A] [answered]
Hi, 

First of all, thank you for your work, Pocketbase is truly amazing!

I would like to modify the response for user records, in particular, I would like the response to have 2 new properties:

- `label`: identical to `username`
- `imageSrc`: identical to `avatar`

The [documentation](https://pocketbase.io/docs/js-event-hooks/#onrecordslistrequest) mentions:

> onRecordsListRequest hook is triggered on each API Records list request. Could be used to validate or modify the response before returning it to the client. 

I was able to do this:

``` javascript
// main.pb.js

/** @type {import('../pb_data/types').onRecordsListRequest} */
onRecordsListRequest((e) => {
  const { normalizeUser } = require(`${__hooks}/utils`)
  e.httpContext.json(200, { items: e.records.map(normalizeUser) })
}, 'users')

/** @type {import('../pb_data/types').onRecordViewRequest} */
onRecordViewRequest((e) => {
  const { normalizeUser } = require(`${__hooks}/utils`)
  e.httpContext.json(200, normalizeUser(e.record))
}, 'users')
```

``` javascript
// utils.js

module.exports = {
  normalizeUser: (user) => {
    // const data = JSON.parse(JSON.stringify(user))
    const data= user.publicExport()

    const result = {
		...data,
		label: data.username,
		image: data.avatar,
    }
    return result
  },
}
```

It *almost* works. The problem is that now the email field in the admin is showing `N/A`

I tried with both `JSON.parse(JSON.stringify(user))` and `user.publicExport()`, same result.

Previously, I tried the approach `record.set("label", record.username)` but `record`  only accepts setting existing fields.

I also tried to spread `record` directly but this just breaks the endpoint:

`const augmentedRecord = {...e.record, label: e.record.username}`

Is there any way to accomplish this?

Thank you!

**Comment by ganigeorgiev** [comment, marked as answer]
You usually don't need to use the spread syntax as `publicExport()` should return an object that you can append directly new elements to it.

> Previously, I tried the approach record.set("label", record.username) but record only accepts setting existing fields.

This should work, but you'll also need to call [`record.withUnknownData(true)`](https://pocketbase.io/jsvm/interfaces/models.Record.html#withUnknownData) to enable serializing the unknown fields.

_Side-note: Serialization hook was requested before and it is in my todo to consider it for a future release._

**Comment by danielres** [comment]
Amazing, thank you!

In case this is useful to anyone else:

Actually, accessing the property name directly didn't work, but it works using the getter:

``` javascript
/** @type {import('../pb_data/types').onRecordViewRequest} */
onRecordViewRequest((e) => {
  e.record.withUnknownData(true)
  e.record.set('label', e.record.get('username'))
  e.record.set('image', e.record.get('avatar'))
}, 'users')
```

**Reply by sadlrxn** [reply]
help me out, please.

```js
onRecordsListRequest((e) => {
  e.records.forEach(r => {
    const discount = r.get("discount")
    if (discount && parseInt(discount) > 0) {
      const discountMultiplier = 1 - (parseInt(discount) / 100)
      r.set("discout_price_usd", r.get("price_usd") * discountMultiplier)
      r.set("discout_price_egp", r.get("price_egp") * discountMultiplier)
    }
  })
  
  // Debug: These values appear correctly in the server logs
  e.records.forEach((r) => {
    console.log(r.get("name"), r.get("discout_price_usd"), r.get("discout_price_egp"))
  })
  
  e.next()
}, "products")
```

thecking, there is no "withUnknownData(true)" available.

**Reply by ganigeorgiev** [reply]
@sadlrxn You are commenting on an old discussion and the above code sample is for PocketBase < 0.22.x.

PocketBase v0.23+ introduced many breaking changes, including renaming `withUnknownData` to `withCusomtData`. Additionally, instead of using the `onRecordsListRequest` hook you may want to check the new [`onRecordEnrich` hook](https://pocketbase.io/docs/js-event-hooks/#onrecordenrich) which will be triggered for each record list, view, realtime event, etc. For example:

```js
onRecordEnrich((e) => {
    const discount = e.record.get("discount")
    if (discount && parseInt(discount) > 0) {
        e.record.withCustomData(true)
        const discountMultiplier = 1 - (parseInt(discount) / 100)
        e.record.set("discout_price_usd", e.record.get("price_usd") * discountMultiplier)
        e.record.set("discout_price_egp", e.record.get("price_egp") * discountMultiplier)
    }

    e.next()
}, "products")
```

If you still need further help, please open a new Q&A discusion but try to provide a little more context about your project and what you are trying to do.

**Reply by sadlrxn** [reply]
onRecordEnrich is not working either: I've tried using onRecordEnrich as an alternative approach to add the calculated discount prices, but it's experiencing the same issue - the new properties don't appear on the client side.
Only affecting single record requests: The current issue impacts not just individual product retrieval but also product list operations. I need a solution that works for both single record fetches and collection list queries.

For reference, I've filed this under issue #6692.

**Reply by ganigeorgiev** [reply]
@sadlrxn Please don't spam. It works but you are missing the `e.record.withCustomData(true)` call.
